### ISSUE TITLE
Enhanced Summarizer API with Recursive Text Splitting

### ISSUE THREAD

**Issue by hemanth** [WICG/writing-assistance-apis#14]
## Problem
Currently, our summarizer API doesn't handle large documents efficiently. When the input text exceeds the model's context window, the API fails to process. Users need to manually split large texts and manage the summarization process themselves, which is error-prone and creates inconsistent results.

## Proposed Enhancement
Add automatic text splitting and recursive summarization capabilities to the API, with progress monitoring through callbacks.

### Key Features

1. **Automatic Document Chunking**
   - Split large documents into manageable chunks automatically
   - Maintain context through overlapping chunks
   - Smart splitting at natural boundaries (sentences/paragraphs)
   - Configurable chunk sizes and overlap amounts

2. **Recursive Summarization**
   - Process chunks recursively for very large documents
   - Combine intermediate summaries intelligently
   - Maintain consistent summarization quality across the document

3. **Progress Monitoring**
   - Callback system to track processing status
   - Monitor individual chunk processing
   - Track intermediate summaries
   - Get completion statistics

### Example Usage

```javascript
const summarizer = await ai.summarizer.create({
  sharedContext: "An article from the Daily Economic News magazine",
  type: "headline",
  length: "short",
  
  // Optional chunking configuration
  chunking: {
    maxChunkSize: 2000,
    overlapSize: 200
  },
  
  // Optional progress callbacks
  callbacks: {
    onChunk: async (chunk, depth) => {
      // Track chunk processing
    },
    onSummary: async (summary, sourceChunk, depth) => {
      // Monitor intermediate summaries
    },
    onComplete: async (finalSummary, stats) => {
      // Handle completion
    }
  }
});
```

## Benefits

1. **Better User Experience**
   - No manual text splitting required
   - Consistent results for documents of any size
   - Progress visibility for long-running summarizations

2. **Improved Summary Quality**
   - Context preservation through chunk overlap
   - Hierarchical summarization for very large documents
   - Consistent summarization approach across chunks

3. **Developer Flexibility**
   - Optional configuration for advanced use cases
   - Progress monitoring for UI updates
   - TypeScript support for better type safety

## Backward Compatibility
The enhanced API maintains full compatibility with the current simple usage pattern:

```javascript
// Simple usage still works
const quickSummary = await ai.summarizer.process(text, {
  type: "headline",
  length: "short"
});
```

## Implementation Considerations

1. **Chunking Strategy**
   - Default chunk size based on model's optimal context window
   - Smart text splitting at sentence/paragraph boundaries
   - Configurable overlap to maintain context

2. **Resource Usage**
   - Manage concurrent chunk processing
   - Consider memory usage for very large documents
   - Optional batch processing for resource constraints

3. **Error Handling**
   - Graceful degradation for partial failures
   - Clear error messages for configuration issues
   - Recovery strategies for failed chunks

**Comment by tomayac** [comment]
@andreban has implemented a client-side solution that essentially does what you describe, modulo the automatic context keeping.

**Comment by andreban** [comment]
Here's a [demo][1] and [source code][2] for the client-side solution. It does support overlapping chunks [here][3]. It uses [langchain.js][4] for chunking text and the Summarizer API for generating summaries, with handwritter recursive summarization.

[1]: https://chrome.dev/web-ai-demos/summary-of-summaries
[2]: https://github.com/GoogleChromeLabs/web-ai-demos/blob/main/summary-of-summaries/
[3]: https://github.com/GoogleChromeLabs/web-ai-demos/blob/main/summary-of-summaries/src/main.ts#L19
[4]: https://js.langchain.com/docs/introduction/

**Comment by hemanth** [comment]
Yes, I was cooking something similar and came across @andreban and felt like this should be part of the API rather! ❤️
